server/protocols/jmap-draft/src/main/java/org/apache/james/jmap/http/DownloadRoutes.java

```diff
@@ -0,0 +1,208 @@
+    }
+
+    private Mono<Void> getTwoParam(HttpServerRequest request, HttpServerResponse response) {
+        String blobId = request.param(BLOB_ID_PATH_PARAM);
+        try {
+            String name = URLDecoder.decode(request.param(NAME_PATH_PARAM), StandardCharsets.UTF_8.toString());
+            DownloadPath downloadPath = DownloadPath.of(blobId, name);
+            return get(request, response, downloadPath);
+        } catch (UnsupportedEncodingException e) {
+            throw new BadRequestException("Wrong url encoding", e);
+        }
+    }
+
+    private Mono<Void> get(HttpServerRequest request, HttpServerResponse response, DownloadPath downloadPath) {
+        return authenticationReactiveFilter.authenticate(request)
+            .flatMap(session -> Mono.from(metricFactory.runPublishingTimerMetric("JMAP-download-get",
+                download(session, downloadPath, response)))
+            .onErrorResume(InternalErrorException.class, e -> handleInternalError(response, e)))
+            .onErrorResume(UnauthorizedException.class, e -> handleAuthenticationFailure(response, e))
+            .subscribeOn(Schedulers.elastic());
+    }
+
+    private Mono<Void> respondAttachmentAccessToken(MailboxSession mailboxSession, DownloadPath downloadPath, HttpServerResponse resp) {
+        String blobId = downloadPath.getBlobId();
+        try {
+            if (!attachmentExists(mailboxSession, blobId)) {
+                return resp.status(NOT_FOUND).send();
+            }
+            AttachmentAccessToken attachmentAccessToken = simpleTokenFactory.generateAttachmentAccessToken(mailboxSession.getUser().asString(), blobId);
+            return resp.header(CONTENT_TYPE, TEXT_PLAIN_CONTENT_TYPE)
+                .status(OK)
+                .sendString(Mono.just(attachmentAccessToken.serialize()))
+                .then();
+        } catch (MailboxException e) {
+            throw new InternalErrorException("Error while asking attachment access token", e);
+        }
+    }
+
+    private boolean attachmentExists(MailboxSession mailboxSession, String blobId) throws MailboxException {
+        try {
+            blobManager.retrieve(BlobId.fromString(blobId), mailboxSession);
+            return true;
+        } catch (BlobNotFoundException e) {
+            return false;
+        }
+    }
+
+    @VisibleForTesting
+    Mono<Void> download(MailboxSession mailboxSession, DownloadPath downloadPath, HttpServerResponse response) {
+        String blobId = downloadPath.getBlobId();
+        try {
+            Blob blob = blobManager.retrieve(BlobId.fromString(blobId), mailboxSession);
+
+            return addContentDispositionHeader(downloadPath.getName(), response)
+                .header("Content-Length", String.valueOf(blob.getSize()))
+                .header(CONTENT_TYPE, blob.getContentType())
+                .status(OK)
+                .send(ReactorUtils.toChunks(blob.getStream(), BUFFER_SIZE)
```

Comment:
the opened stream need to be closed

Reply:
I get your point but I'm struggling doing it... How would you do it?

Reply:
Usually `Mono.usingWhen()`  I'm sure @LanKhuat is the master of using it

Reply:
I tried, couldn't make it work